src/transformers/models/aria/modular_aria.py

```diff
@@ -0,0 +1,1378 @@
+import inspect
+import os
+from typing import Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from torch import nn
+from torch.nn.init import trunc_normal_
+
+from ...activations import ACT2FN
+from ...configuration_utils import PretrainedConfig
+from ...feature_extraction_utils import BatchFeature
+from ...generation import GenerationMixin
+from ...image_processing_utils import BaseImageProcessor, select_best_resolution
+from ...image_transforms import (
+    convert_to_rgb,
+    resize,
+    to_channel_dimension_format,
+    pad,
+)
+from ...image_utils import (
+    ChannelDimension,
+    ImageInput,
+    PILImageResampling,
+    get_image_size,
+    to_numpy_array,
+)
+from ...modeling_utils import PreTrainedModel
+from ...processing_utils import ProcessorMixin
+from ...tokenization_utils import (
+    PaddingStrategy,
+    PreTokenizedInput,
+    TensorType,
+    TextInput,
+    TruncationStrategy,
+)
+from ...utils import (
+    logging,
+)
+from ...utils.import_utils import is_vision_available
+from ..auto import CONFIG_MAPPING, AutoModel, AutoModelForCausalLM, AutoTokenizer
+from ..llama.configuration_llama import LlamaConfig
+from ..llama.modeling_llama import (
+    LLAMA_ATTENTION_CLASSES,
+    LlamaDecoderLayer,
+    LlamaForCausalLM,
+    LlamaMLP,
+    LlamaModel,
+    LlamaPreTrainedModel,
+    LlamaRMSNorm,
+)
+from ..llava.modeling_llava import LlavaCausalLMOutputWithPast
+
+
+logger = logging.get_logger(__name__)
+
+
+if is_vision_available():
+    from PIL import Image, ImageOps
+
+
+def sequential_gemm(input, weight, tokens_per_expert):
+    """
+    Compute the matrix multiplication (GEMM) for each expert sequentially. This approach is computationally inefficient, especially when dealing with a large number of experts.
+
+    Args:
+        input (torch.Tensor): Input tensor of shape (num_tokens, in_features).
+        weight (torch.Tensor): Weight tensor of shape (num_experts, in_features, out_features).
+        tokens_per_expert (torch.Tensor): Number of tokens assigned to each expert.
+
+    Returns:
+        torch.Tensor: Output tensor of shape (num_tokens, out_features).
+    """
+    num_tokens = input.shape[0]
+    out_features = weight.shape[-1]
+    output = torch.zeros(num_tokens, out_features, dtype=input.dtype, device=input.device)
+
+    cumsum_num_tokens = torch.cumsum(tokens_per_expert, dim=0)
+    # Insert zero at the begining for offset index's convenience
+    zero_tensor = torch.zeros(1, dtype=torch.long, device=cumsum_num_tokens.device)
+    cumsum_num_tokens = torch.cat((zero_tensor, cumsum_num_tokens))
+
+    for expert_num in range(weight.shape[0]):
+        start = cumsum_num_tokens[expert_num]
+        end = cumsum_num_tokens[expert_num + 1]
+        tokens = input[start:end]
+
+        out = torch.matmul(tokens, weight[expert_num])
+        output[start:end] = out
+    return output
```

Comment:
@molbap I looked it up, but alas it won't work (at least not the way I tried):
the reason for this is that in the loop, the `tokens` matrix has a different size for each expert (since a different number of input tokens are attributed to each expert). And `torch.bmm` requires 3d tensors, which requires the same length for all `tokens` matrices before stacking. We could pad to the max number of tokens and then do the batched multiply, but that would create a memory overhead.